applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp

```diff
@@ -272,69 +274,176 @@ void DistanceModificationProcess::ModifyDistance() {
-                for (unsigned int i_node = 0; i_node < r_elem_dist.size(); ++i_node){
+                for (size_t i_node = 0; i_node < r_elem_dist.size(); ++i_node){
                     if (std::abs(r_elem_dist(i_node)) < tol_d){
-                        if (!is_saved){
-                            aux_modified_distances_ids.push_back(it_elem->Id());
-                            aux_modified_elemental_distances.push_back(r_elem_dist);
-                        }
                         r_elem_dist(i_node) = r_elem_dist(i_node) > 0.0 ? tol_d : -tol_d;
                     }
                 }
             }
+        }
+    } else {
+        // Case in which the original distance needs to be kept to track the interface (e.g. FSI)
 
-            // Save the auxiliar chunk arrays
-            #pragma omp critical
+        const size_t num_chunks = 2 * ParallelUtilities::GetNumThreads();
+        OpenMPUtils::PartitionVector partition_vec;
+        OpenMPUtils::DivideInPartitions(n_elems,num_chunks,partition_vec);
+
+        // If user has provided flag to calculate elemental edge distances of extrapolated geometry,
+        // modify them according to elemental distances in order to stay consistent with element splitting
+        if (n_edges_extrapolated > 0) {
+            #pragma omp parallel for
+            for (size_t i_chunk = 0; i_chunk < num_chunks; ++i_chunk)
             {
-                mModifiedDistancesIDs.insert(mModifiedDistancesIDs.end(),aux_modified_distances_ids.begin(),aux_modified_distances_ids.end());
-                mModifiedElementalDistancesValues.insert(mModifiedElementalDistancesValues.end(),aux_modified_elemental_distances.begin(),aux_modified_elemental_distances.end());
+                auto elems_begin = r_elems.begin() + partition_vec[i_chunk];
+                auto elems_end = r_elems.begin() + partition_vec[i_chunk + 1];
+
+                // Auxiliar chunk arrays
+                std::vector<size_t> aux_modified_distances_ids;
+                std::vector<Vector> aux_modified_elemental_distances;
+                std::vector<size_t> aux_modified_edge_dist_extra_ids;
+                std::vector<Vector> aux_modified_edge_dist_extra;
+
+                for (auto it_elem = elems_begin; it_elem < elems_end; ++it_elem) {
+                    // Compute the distance tolerance
+                    const double tol_d = mDistanceThreshold * (it_elem->GetGeometry()).Length();
+
+                    bool is_saved = false;
+                    Vector &r_elem_dist = it_elem->GetValue(ELEMENTAL_DISTANCES);
+                    for (size_t i_node = 0; i_node < r_elem_dist.size(); ++i_node){
+                        if (std::abs(r_elem_dist(i_node)) < tol_d){
+                            if (!is_saved){
+                                aux_modified_distances_ids.push_back(it_elem->Id());
+                                aux_modified_elemental_distances.push_back(r_elem_dist);
+                                is_saved = true;
```

Comment:
I added this line, because it seemed quite useful to save time and storage and it seemed to be the purpose for creating the `is_saved` variable in the first place. Am I wrong?

Reply:
Yes. I think this was the original idea. However, we've never used the recovered distances (we always stay with the modified ones). Having said this, my suggestion would be to directly remove this part of the code.